### PR TITLE
Cleaning up ReductionOp

### PR DESCRIPTION
**Pull request by naoyam**
Cleans up ReductionOp and GridReduction. GridReduction inherits ReductionOp. A grid reduction is now
represented with just a single GridReduction expression (previously it was with ReductionOp AND GridReduction). This simplifies the relevant code in lower_index as we now need to just create GridReduction for a grid reduction call.

The runtime gridReduction code is extended to do block reduction as well. This simplifies code generation as it does not need to generate a separate block-reduction call.

I'll do similar cleanup for Welford once this PR is approved.

### CODE REVIEW

**Comment by naoyam** [comment]
Rebased onto TOT

File: torch/csrc/jit/codegen/cuda/kernel.cpp [hunk 1]
```diff
@@ -122,11 +122,7 @@ class KernelIrScanner : private IrVisitor {
 
   void handle(GridReduction* grid_reduction) final {
     summary_.has_grid_reductions = true;
-    const auto dom = grid_reduction->reduction_op()
-                         ->out()
-                         ->as<TensorIndex>()
-                         ->view()
-                         ->domain();
+    const auto dom = grid_reduction->out()->as<TensorIndex>()->view()->domain();
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/kernel.cpp]: Please use `lower_utils.cpp::getTvOutput(...)`

**Review by csarofeen — APPROVED** [review]
Scanned it quickly, don't see any red flags. Just one minor comment.

Is the goal of the cleanup to eventually get rid of the non-generic functions and to always use the mega-grid-enabled function?

**Comment by naoyam** [comment]
I'm thinking about adding a grouped reduction class to represent horizontally fused reductions. I wanted to clean the existing reduction classes first.

I'm not planning to get rid of the old grid reduction kernel (yet). It's been stable, and I think it's good to have it as a fallback option.

**Pull request merged**